extensions/aerosol_optics/mo_aerosol_optics.F90

```diff
@@ -0,0 +1,681 @@
+                 this%aero_rh)
+    end if
+
+    ! Lookup table aerosol optics coefficients
+    if(allocated(this%aero_dust_ext)) then
+
+      deallocate(this%aero_dust_ext, this%aero_dust_ssa, this%aero_dust_asy, &
+                 this%aero_salt_ext, this%aero_salt_ssa, this%aero_salt_asy, &
+                 this%aero_sulf_ext, this%aero_sulf_ssa, this%aero_sulf_asy, &
+                 this%aero_bcar_ext, this%aero_bcar_ssa, this%aero_bcar_asy, &
+                 this%aero_bcar_rh_ext, this%aero_bcar_rh_ssa, this%aero_bcar_rh_asy, &
+                 this%aero_ocar_ext, this%aero_ocar_ssa, this%aero_ocar_asy, &
+                 this%aero_ocar_rh_ext, this%aero_ocar_rh_ssa, this%aero_ocar_rh_asy)
+    end if
+
+  end subroutine finalize
+  ! ------------------------------------------------------------------------------
+  !
+  ! Derive aerosol optical properties from provided aerosol input properties
+  !
+  ! ------------------------------------------------------------------------------
+  !
+  ! Compute single-scattering properties
+  !
+  function aerosol_optics(this, aero_type, aero_size, aero_mass, relhum, &
+                          optical_props) result(error_msg)
+
+    class(ty_aerosol_optics), &
+              intent(in  ) :: this
+    integer,  intent(in  ) :: aero_type(:,:)   ! MERRA2/GOCART aerosol type
+    real(wp), intent(in  ) :: aero_size(:,:)   ! aerosol size for dust and sea-salt
+    real(wp), intent(in  ) :: aero_mass(:,:)   ! aerosol mass column (g/m2)
+    real(wp), intent(in  ) :: relhum(:,:)      ! relative humidity (fraction, 0-1)
+
+    class(ty_optical_props_arry), &
+              intent(inout) :: optical_props
+                                               ! Dimensions: (ncol,nlay,nbnd)
+
+    character(len=128)      :: error_msg
+    ! ------- Local -------
+    logical(wl), dimension(size(aero_type,1), size(aero_type,2)) :: aeromsk
+    real(wp),    dimension(size(aero_type,1), size(aero_type,2), size(this%aero_dust_ext,2)) :: &
+!    real(wp),    dimension(size(aero_type,1), size(aero_type,2), this%get_nband()) :: &
+                 atau, ataussa, ataussag
+                ! ltau, ltaussa, ltaussag, itau, itaussa, itaussag
+                ! Optical properties: tau, tau*ssa, tau*ssa*g
+    integer  :: ncol, nlay, nbnd, nbin, nrh
+    integer  :: icol, ilay, ibnd, ibin
+    ! scalars for total tau, tau*ssa
+    real(wp) :: tau, taussa
+
+    ! ----------------------------------------
+    !
+    ! Error checking
+    !
+    ! ----------------------------------------
+
+    error_msg = ''
+    if(.not.(allocated(this%aero_dust_ext))) then
+      error_msg = 'aerosol optics: no data has been initialized'
+      return
+    end if
+
+    ncol = size(aero_type,1)
+    nlay = size(aero_type,2)
+    nbin = size(this%aero_bin_lo,1)
+    nrh = size(this%aero_rh,1)
+!    nbnd = this%get_nband()
+    nbnd = size(this%aero_dust_ext,2)
+
+    write(61,*) 'Aerosol Optics: checking arrays...'
```

Comment:
@mjiacono Please remove write/print statements. Classes should not input or output.

Reply:
All write statements removed.